### PR TITLE
Bug/vspc 218

### PR DESCRIPTION
**Pull request by swetalina-prusty25**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

when I’m on a module page on the staff side, there needs to be a list of all spaces that link to that module.

...Put description here...
On staff side: module page needs to show from which spaces it is linked
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 1]
```diff
@@ -3,17 +3,29 @@
...
 
 @Repository
 @JaversSpringDataAuditable
 public interface ModuleLinkRepository extends PagingAndSortingRepository<ModuleLink, String> {
+
+	
+    List<ModuleLink> findByModuleId(IModule module);
+       
+    @Query("select m from ModuleLink m where m.module.id = ?1")
+    List<ModuleLink> getModuleLinks(String moduleId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:24]: can't you just use the automatic query generation here? It's just a getByModule isn't it?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 2]
```diff
@@ -63,28 +69,28 @@ protected IModuleLink createLinkObject(String title, String id) {
...
+    protected IModuleLink getLink(String moduleLinkID) {
+        Optional <ModuleLink> moduleLink = moduleLinkRepo.findById(moduleLinkID);
+        if (moduleLink.isPresent()) {
             return moduleLink.get();
         }
         return null;
     }
 
-    @Override 
+    @Override
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:93]: please avoid reformatting all the above if there are no changes.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 3]
```diff
@@ -104,4 +110,14 @@ protected void deleteLinkRepo(IModuleLink link) {
         moduleLinkRepo.delete((ModuleLink) link);
     }
 
+    @Override
+    public HashSet <ISpace> findModuleLinksFromModuleId(String moduleId) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:114]: please fix your ide settings, we don't do blanks before `<`

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 4]
```diff
@@ -3,17 +3,29 @@
...
+    
+    @Query("select s.space.id from ModuleLink s where s.id = ?1")
+    String getSpaceIdFromModuleLink(String link);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:27]: what is this method for? why not just getting the module link and then do `link.getSpace().getId()`?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 3 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:114]: this method is called find**ModuleLinks**FromModuleId, why is it returning spaces?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 5]
```diff
@@ -104,4 +110,14 @@ protected void deleteLinkRepo(IModuleLink link) {
...
+        List <ModuleLink> moduleLinks = moduleLinkRepo.getModuleLinks(moduleId);
+        HashSet <ISpace> spaces = new HashSet <>();
+        for (ModuleLink link: moduleLinks) {
+            spaces.add(link.getSpace());
+        }
+        return spaces;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:120]: the above can be a one-liner using the stream api

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java [hunk 6]
```diff
@@ -1,29 +1,47 @@
...
     public String showModule(@PathVariable String id, Model model) {
         IModule module = moduleManager.getModule(id);
         model.addAttribute("module", module);
         model.addAttribute("slides", moduleManager.getModuleSlides(id));
         model.addAttribute("sequences", moduleManager.getModuleSequences(id));
         model.addAttribute("moduleStatus", module.getModuleStatus());
+        HashSet <ISpace> spaces = moduleLink.findModuleLinksFromModuleId(id);
+        if (!spaces.isEmpty()) {
+            
+            model.addAttribute("spacesList", spaces);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:42]: it doesn't hurt adding an empty list here, does it? is the if statement really necessary?

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 4 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:27]: also, is this method actually used?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java (hunk 6 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:42]: If statement is not necessarily required here, so fixed this.

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 4 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:27]: This getSpaceIdFromModuleLink() has not been used anywhere, i have added these for testing purpose. so fixed this issue.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 3 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:114]: Fixed the ide setting and renamed the method name to findSpaceLinksFromModuleId

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 1 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:24]: used getByModule

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 5 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:120]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 7]
```diff
@@ -3,17 +3,25 @@
...
 
 @Repository
 @JaversSpringDataAuditable
 public interface ModuleLinkRepository extends PagingAndSortingRepository<ModuleLink, String> {
+
+	
+    List<ModuleLink> findByModuleId(IModule module);
+          
+    @Query("select m from ModuleLink m where m.module.id = ?1")
+    List<ModuleLink>getByModule(String moduleId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:24]: These two methods seem to do the same. Why are they both here? Just the first one should suffice (and you should be able to give it the module id).

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 8]
```diff
@@ -45,7 +51,7 @@ public class ModuleLinkManager extends LinkManager<IModuleLink,IModule,IModuleLi
     private ModuleLinkDisplayRepository moduleLinkDisplayRepo;
 
     @Override
-    public List<IModuleLinkDisplay> getLinkDisplays(String spaceId) {
+    public List <IModuleLinkDisplay> getLinkDisplays(String spaceId) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:54]: please fix your settings, there should be no blank here.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 5 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:118]: 116-118 can be reduced to one line (using `collect`).

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 5 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:118]: used (collect). to reduce lines

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 7 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:24]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 8 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:54]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 9]
```diff
@@ -104,4 +111,10 @@ protected void deleteLinkRepo(IModuleLink link) {
         moduleLinkRepo.delete((ModuleLink) link);
     }
 
+    @Override
+    public HashSet<ISpace> findSpaceLinksFromModuleId(String moduleId) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:115]: this method returns a list of spaces, why is is called "findSpaceLinks..."?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 9 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:115]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 10]
```diff
@@ -3,18 +3,23 @@
...
+import edu.asu.diging.vspace.core.model.IModule;
 import edu.asu.diging.vspace.core.model.impl.ModuleLink;
 import edu.asu.diging.vspace.core.model.IModuleLink;
 
 @Repository
 @JaversSpringDataAuditable
 public interface ModuleLinkRepository extends PagingAndSortingRepository<ModuleLink, String> {
-    
+
+    @Query("select m from ModuleLink m where m.module.id = ?1")
```
[the remark this answers is not in the document]
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:20]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 11]
```diff
@@ -3,18 +3,23 @@
...
+import edu.asu.diging.vspace.core.model.IModule;
 import edu.asu.diging.vspace.core.model.impl.ModuleLink;
 import edu.asu.diging.vspace.core.model.IModuleLink;
 
 @Repository
 @JaversSpringDataAuditable
 public interface ModuleLinkRepository extends PagingAndSortingRepository<ModuleLink, String> {
+
     
+    List<ModuleLink> findByModuleId(String moduleId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:21]: how is this method different from the next one?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java [hunk 12]
```diff
@@ -104,4 +111,10 @@ protected void deleteLinkRepo(IModuleLink link) {
         moduleLinkRepo.delete((ModuleLink) link);
     }
 
+    @Override
+    public HashSet<ISpace> findSpaceListFromModuleId(String moduleId) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:115]: method should just return `Set<ISpace>`

File: vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html [hunk 13]
```diff
@@ -339,6 +339,19 @@ <h4>All Slides</h4>
                 </div>
             </div>              
         </div>
+        <div class="tab sequences" style="margin-top: 10px;">
+            <div style="display: flex; justify-content: space-between">
+               <h4>List of Spaces Linked to a Module</h4>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:344]: Should be "Spaces linking to this Module"

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java (hunk 12 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java:115]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java [hunk 14]
```diff
@@ -1,9 +1,16 @@
...
 import edu.asu.diging.vspace.core.model.IModule;
 import edu.asu.diging.vspace.core.model.IModuleLink;
+import edu.asu.diging.vspace.core.model.ISpace;
 import edu.asu.diging.vspace.core.model.display.IModuleLinkDisplay;
+import edu.asu.diging.vspace.core.model.impl.ModuleLink;
+
+public interface IModuleLinkManager extends ILinkManager <IModuleLink, IModule, IModuleLinkDisplay> {
 
-public interface IModuleLinkManager extends ILinkManager<IModuleLink,IModule,IModuleLinkDisplay>{
+    HashSet <ISpace> findSpaceListFromModuleId(String moduleId);
```
[the remark this answers is not in the document]
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java:14]: fixed

File: vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html (hunk 13 above)
Comment (swetalina-prusty25) [reply vspace/src/main/webapp/WEB-INF/views/staff/modules/module.html:344]: fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 11 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:21]: This method is not different from the next one. so updated the code according to the existed method.

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java [hunk 15]
```diff
@@ -3,18 +3,20 @@
 import java.util.List;
 
 import org.javers.spring.annotation.JaversSpringDataAuditable;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.PagingAndSortingRepository;
 import org.springframework.stereotype.Repository;
 import org.springframework.data.jpa.repository.Modifying;
 import org.springframework.data.jpa.repository.Query;
 
+import edu.asu.diging.vspace.core.model.IModule;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:12]: it doesn't look like these imports are used anywhere?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java [hunk 16]
```diff
@@ -1,29 +1,45 @@
...
+
     @RequestMapping(STAFF_MODULE_PATH+"{id}")
     public String showModule(@PathVariable String id, Model model) {
         IModule module = moduleManager.getModule(id);
         model.addAttribute("module", module);
         model.addAttribute("slides", moduleManager.getModuleSlides(id));
         model.addAttribute("sequences", moduleManager.getModuleSequences(id));
         model.addAttribute("moduleStatus", module.getModuleStatus());
+        Set<ISpace> spaces = moduleLink.findSpaceListFromModuleId(id);
+        model.addAttribute("spacesList", spaces);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:42]: merge with line 41

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
There are compilation errors
```
[ERROR] symbol:   class ModuleNotFoundException
[ERROR] location: package edu.asu.diging.vspace.web.exception
[ERROR] /var/lib/jenkins/workspace/VSpace_deploy_to_review_on_pr/vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:[21,43] cannot find symbol
[ERROR] symbol:   class SpaceNotFoundException
[ERROR] location: package edu.asu.diging.vspace.web.exception
[ERROR] -> [Help 1]
[ERROR] 
```

**Comment by swetalina-prusty25** [comment]
i have removed class ModuleNotFoundException, class SpaceNotFoundException packages from the controller. As with or without those exception packages i am able to compile the code in my locals.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java (hunk 16 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:42]: Fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java (hunk 15 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/ModuleLinkRepository.java:12]: Fixed

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java [hunk 17]
```diff
@@ -1,9 +1,14 @@
...
 
+import java.util.Set;
+
 import edu.asu.diging.vspace.core.model.IModule;
 import edu.asu.diging.vspace.core.model.IModuleLink;
+import edu.asu.diging.vspace.core.model.ISpace;
 import edu.asu.diging.vspace.core.model.display.IModuleLinkDisplay;
 
-public interface IModuleLinkManager extends ILinkManager<IModuleLink,IModule,IModuleLinkDisplay>{
+public interface IModuleLinkManager extends ILinkManager <IModuleLink, IModule, IModuleLinkDisplay> {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java:10]: please fix your ide settings; there should be no blanks before `<`

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
The spaces in the list should be links to the space page.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java [hunk 18]
```diff
@@ -5,25 +5,29 @@
...
 @Controller
 public class ModuleController {
 
     public static final String STAFF_MODULE_PATH = "/staff/module/";
+
     @Autowired
     private IModuleManager moduleManager;
 
+    @Autowired
+    private IModuleLinkManager moduleLink;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:21]: This is very confusing. The object is not a module link but a module link manager. Right now line 30 implies that a module link has a method to get a space list from a module id.

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java [hunk 19]
```diff
@@ -1,9 +1,14 @@
...
+
 import edu.asu.diging.vspace.core.model.IModule;
 import edu.asu.diging.vspace.core.model.IModuleLink;
+import edu.asu.diging.vspace.core.model.ISpace;
 import edu.asu.diging.vspace.core.model.display.IModuleLinkDisplay;
 
-public interface IModuleLinkManager extends ILinkManager<IModuleLink,IModule,IModuleLinkDisplay>{
+public interface IModuleLinkManager extends ILinkManager<IModuleLink, IModule, IModuleLinkDisplay> {
+
+    Set<ISpace> findSpaceListFromModuleId(String moduleId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java:12]: needs javadoc

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
The icons for the spaces in the list should have the space icon not the module icon.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java (hunk 18 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ModuleController.java:21]: Changed the object name to moduleLinkManager

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java (hunk 19 above)
Comment (swetalina-prusty25) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/IModuleLinkManager.java:12]: Added javadov

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Pull request merged**